index.js

```diff
@@ -436,6 +436,105 @@ DumbPipe.registerOpener("camera", function(message, sender) {
   };
 });
 
+var LocationProvider = {};
+LocationProvider.AVAILABLE = 1;
+LocationProvider.TEMPORARILY_UNAVAILABLE = 1;
+LocationProvider.OUT_OF_SERVICE = 1;
+LocationProvider.state = LocationProvider.OUT_OF_SERVICE;
+LocationProvider.position = {
+  timestamp: 0,
+  latitude: 0,
+  longitude: 0,
+  altitude: NaN,
+  horizontalAccuracy: NaN,
+  verticalAccuracy: NaN,
+  speed: NaN,
+  heading: NaN
+};
+
+DumbPipe.registerOpener("locationprovider", function(message, sender) {
+  var id = -1;
+  var waitForData = false;
+
+  // Used to stop watching geolocation if it hasn't been accessed for a given
+  // time period.
+  var clearWatchInterval = -1;
+  var lastVisited = 0;
+
+  function watchPosition() {
+    lastVisited = Date.now();
+    if (id !== -1) {
+      return;
+    }
+    LocationProvider.state = LocationProvider.TEMPORARILY_UNAVAILABLE;
+    id = navigator.geolocation.watchPosition(function(pos) {
+        LocationProvider.state = LocationProvider.AVAILABLE;
+        LocationProvider.position.timestamp = pos.timestamp;
+
+        var c = pos.coords;
+        LocationProvider.position.latitude = c.latitude;
+        LocationProvider.position.longitude = c.longitude;
+        LocationProvider.position.altitude =
+          c.altitude !== null ? c.altitude : NaN;
+        LocationProvider.position.horizontalAccuracy =
+          c.accuracy !== null ? c.accuracy : NaN;
+        LocationProvider.position.verticalAccuracy =
+          c.altitudeAccuracy !== null ? c.altitudeAccuracy : NaN;
+        LocationProvider.position.speed = c.speed !== null ? c.speed : NaN;
+        LocationProvider.position.course = c.heading !== null ? c.heading : NaN;
+        sendData();
+    }, function(err) {
+        console.log("locationprovider error: " + err);
+        LocationProvider.state = LocationProvider.OUT_OF_SERVICE;
+        clearWatch();
+        sendData();
+    });
+    clearWatchInterval = setInterval(function() {
+      // Stop watching after being inactive for 30s to save power.
+      if (Date.now() - lastVisited > 30 * 1000) {
+        clearWatch();
+      }
+    }, 10000);
+  }
+
+  function clearWatch() {
+    if (id === -1) {
+      return;
+    }
+    clearInterval(clearWatchInterval);
+    clearWatchInterval = -1;
+    navigator.geolocation.clearWatch(id);
+    id = -1;
+  }
+
+  function sendData() {
+    if (!waitForData) {
+      return;
+    }
+    sender({
+      type: "data",
+      position: LocationProvider.position,
+      state: LocationProvider.state
+    });
+    waitForData = false;
+  }
+
+  watchPosition();
+
+  return function(message) {
+    switch (message.type) {
+      case "close":
+        clearWatch();
+        break;
+
+      case "requestData":
+        waitForData = true;
+        watchPosition();
+        break;
+    }
+  };
+});
```

Comment:
In a followup, we should start preprocessing this code out if the Java API is disabled, either by preprocessing this file or by moving this code into a separate script that we include conditionally.